src/graphinfo.jl

```diff
@@ -0,0 +1,254 @@
+    GraphInfo
+
+Record the state of the model as a struct of NamedTuples, all
+sharing the same key values, namely, those of the model parameters.
+`value` should store the initial/current value of the parameters.
+`input` stores a tuple of inputs for a given node. `eval` are the
+anonymous functions associated with each node. These might typically
+be either deterministic values or some distribution, but could an 
+arbitrary julia program. `kind` is a tuple of symbols indicating
+whether the node is a logical or stochastic node. Additionally, the 
+adjacency matrix and topologically ordered vertex list and stored.
+
+GraphInfo is instantiated using the `Model` constctor. 
+"""
+
+struct GraphInfo{T} <: AbstractModelTrace
+    value::NamedTuple{T}
+    input::NamedTuple{T}
+    eval::NamedTuple{T}
+    kind::NamedTuple{T}
+    A::SparseMatrixCSC
+    sorted_vertices::Vector{Symbol}
+end
+
+"""
+    Model(;kwargs...)
+
+`Model` type constructor that takes in named arguments for 
+nodes and returns a `Model`. Nodes are pairs of variable names
+and tuples containing default value, an eval function 
+and node type. The inputs of each node are inferred from 
+their anonymous functions. The returned object has a type 
+GraphInfo{(sorted_vertices...)}.
+
+# Examples
+```jl-doctest
+julia> using AbstractPPL
+
+julia> Model(
+               s2 = (0.0, () -> InverseGamma(2.0,3.0), :Stochastic), 
+               μ = (1.0, () -> 1.0, :Logical), 
+               y = (0.0, (μ, s2) -> MvNormal(μ, sqrt(s2)), :Stochastic)
+           )
+Nodes: 
+μ = (value = 1.0, input = (), eval = var"#6#9"(), kind = :Logical)
+s2 = (value = 0.0, input = (), eval = var"#5#8"(), kind = :Stochastic)
+y = (value = 0.0, input = (:μ, :s2), eval = var"#7#10"(), kind = :Stochastic)
+```
+"""
+
+struct Model{T}
+    g::GraphInfo{T}
+end
+
+function Model(;kwargs...)
+    vals = getvals(NamedTuple(kwargs))
+    args = [argnames(f) for f in vals[2]]
+    A, sorted_vertices = DAG(NamedTuple{keys(kwargs)}(args))    
+    modelinputs = NamedTuple{Tuple(sorted_vertices)}.([vals[1], Tuple.(args), vals[2], vals[3]])
+    Model(GraphInfo(modelinputs..., A, sorted_vertices))
+end
+
+
+"""
+    DAG(inputs)
+
+Function taking in a NamedTuple containing the inputs to each node 
+and returns the implied adjacency matrix and topologically ordered 
+vertex list.
+"""
+function DAG(inputs)
+    input_names = Symbol[keys(inputs)...]
+    println(inputs)
+    A = adjacency_matrix(inputs) 
+    sorted_vertices = topological_sort_by_dfs(A)
+    sorted_A = permute(A, collect(1:length(inputs)), sorted_vertices)
+    sorted_A, input_names[sorted_vertices]
+end
+
+"""
+    getvals(nt::NamedTuple{T})
+
+Takes in the arguments to Model(;kwargs...) as a NamedTuple and 
+reorders into a tuple of tuples each containing either of value, 
+input, eval and kind, as required by the GraphInfo type. 
+"""
+@generated function getvals(nt::NamedTuple{T}) where T
+    values = [:(nt[$i][$j]) for i in 1:length(T), j in 1:3]
```

Comment:
Perhaps consider adding some sanity checking on `nt`, similar to those for `Model`?

Reply:
I've added checks in the `Model` constructor to make sure the order and types are as expected. If these are correct then `getvals` should work as expected.